File: src/Calculator/Controls/CalculationResultAutomationPeer.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "pch.h"
#include "CalculationResultAutomationPeer.h"

using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Automation::Peers;

namespace CalculatorApp::Controls
{
    CalculationResultAutomationPeer::CalculationResultAutomationPeer(FrameworkElement ^ owner)
        : FrameworkElementAutomationPeer(owner)
    {
    }

    AutomationControlType CalculationResultAutomationPeer::GetAutomationControlTypeCore()
    {
        return AutomationControlType::Text;
    }

    Platform::Object ^ CalculationResultAutomationPeer::GetPatternCore(PatternInterface pattern)
    {
        if (pattern == PatternInterface::Invoke)
        {
            return this;
        }

        return FrameworkElementAutomationPeer::GetPatternCore(pattern);
    }

    void CalculationResultAutomationPeer::Invoke()
    {
        auto owner = static_cast<CalculationResult ^>(this->Owner);
        owner->ProgrammaticSelect();
    }
}

File: src/Calculator/Controls/CalculationResultAutomationPeer.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using Windows.UI.Xaml;
using Windows.UI.Xaml.Automation.Peers;

namespace CalculatorApp
{
    namespace Controls
    {
        public sealed class CalculationResultAutomationPeer : Windows.UI.Xaml.Automation.Peers.FrameworkElementAutomationPeer,
                                                              Windows.UI.Xaml.Automation.Provider.IInvokeProvider
        {
            public CalculationResultAutomationPeer(FrameworkElement owner) : base(owner)
            {
            }

            protected override AutomationControlType GetAutomationControlTypeCore()
            {
                return AutomationControlType.Text;
            }

            protected override object GetPatternCore(PatternInterface pattern)
            {
                return pattern == PatternInterface.Invoke ? this : base.GetPatternCore(pattern);
            }

            public void Invoke()
            {
                var owner = (CalculationResult)this.Owner;
                owner.ProgrammaticSelect();
            }
        }
    }
}
